docs/the-new-architecture/turbo-modules-with-swift.md

```diff
@@ -0,0 +1,165 @@
+# iOS - Using Swift in Your Native Modules
+
+Swift is the official and default language for developing native application on iOS.
+
+The core of React Native is mainly written in C++ and the interoperability between Swift and C++ is not great, despite the [interoperability layer](https://www.swift.org/documentation/cxx-interop/) Apple developed and released in 2023.
+
+In this guide, we will explore how you can write your Native Modules using Swift. It won't be a pure Swift implementation due to the incompatibilities between the languages, but it minimizes the amount of Objective-C++ glue code that you need to write. If you are migrating an existing Native Modules from the legacy architecture to the New Architecture, this approach should allow you to reuse most of the code.
+
+This guide starts from the iOS implementation of the [Native Module](/docs/next/turbo-native-modules-introduction) guide.
+Make sure to be familiar with that guide before diving into this one, potentially implementing the example in the guide.
+
+## The Adapter pattern
+
+The goal we have is to implement all our business logic using a Swift module and have a thin glue layer in Objective-C++ that is able to connect the app with the Swift implementation.
+
+We can achieve this by leveraging the [Adapter](https://refactoring.guru/design-patterns/adapter) design pattern, where our _adaptee_ is the Swift Module and the Objective-C++ layer is the _adapter_.
+
+The _adapter_ is created by React Native and it keep a reference to the _adaptee_, handling its lifecycle, and it forwards to the _adaptee_ all the methods invocation.
+
+![Adapter Pattern](/docs/assets/AdapterPatternSwiftModules.png)
```

Comment:
nit: not a huge fan we have all sorts of different diagrams in the website :( But I guess better than nothing